modules/javafx.graphics/src/main/java/com/sun/javafx/sg/prism/NGShape3D.java

```diff
@@ -179,61 +171,95 @@ private void renderMeshView(Graphics g) {
-                            0, 0, 0,                 // position is irrelevant
-                            rL, gL, bL, 1.0f,
-                            1, 0, 0,                 // attenuation is irrelevant
-                            0,
-                            Float.POSITIVE_INFINITY, // range is irrelevant
-                            (float) light.getDirection().getX(),
-                            (float) light.getDirection().getY(),
-                            (float) light.getDirection().getZ(),
-                            0, 0, 0);                // spotlight factors are irrelevant
-                } else if (lightBase instanceof NGAmbientLight) {
-                    // Accumulate ambient lights
-                    ambientRed   += rL;
-                    ambientGreen += gL;
-                    ambientBlue  += bL;
+                    addDirectionalLight((NGDirectionalLight) lightBase, lightIndex++, rL, gL, bL);
                 }
             }
-            ambientRed = Utils.clamp(0, ambientRed, 1);
-            ambientGreen = Utils.clamp(0, ambientGreen, 1);
-            ambientBlue = Utils.clamp(0, ambientBlue, 1);
-            meshView.setAmbientLight(ambientRed, ambientGreen, ambientBlue);
+            if (ambientRed != 0.0f || ambientGreen != 0.0f || ambientBlue != 0.0f) {
+                ambientRed = Utils.clamp(0, ambientRed, 1);
+                ambientGreen = Utils.clamp(0, ambientGreen, 1);
+                ambientBlue = Utils.clamp(0, ambientBlue, 1);
+                meshView.setAmbientLight(ambientRed, ambientGreen, ambientBlue);
+            }
         }
         // TODO: 3D Required for D3D implementation of lights, which is limited to 3
 
         while (lightIndex < 3) { // Reset any previously set lights
-            meshView.setLight(lightIndex++,
-                    0, 0, 0,    // x y z
-                    0, 0, 0, 0, // r g b w
-                    1, 0, 0, 1, 0, // ca la qa isAttenuated maxRange
-                    0, 0, 1,    // dirX Y Z
-                    0, 0, 0);   // inner outer falloff
+            resetLight(lightIndex++);
         }
+    }
 
-        meshView.render(g);
+    private boolean noLights(NGLightBase[] lights) {
+        return lights == null || lights[0] == null;
+    }
+
+    /**
+     * If no lights are in the scene, add a default white point light at the camera's. The light uses the default
+     * attenuation parameters, which means that it is not attenuated.
+     */
+    private void createDefaultLight(Graphics g) {
+        meshView.setAmbientLight(0.0f, 0.0f, 0.0f);
+        Vec3d cameraPos = g.getCameraNoClone().getPositionInWorld(null);
+        var direction = NGPointLight.getSimulatedDirection();
+        meshView.setLight(0,
+                (float) cameraPos.x, (float) cameraPos.y, (float) cameraPos.z,
+                1.0f, 1.0f, 1.0f, 1.0f,
+                NGPointLight.getDefaultCa(), NGPointLight.getDefaultLa(), NGPointLight.getDefaultQa(), 0,
```

Comment:
Yes, in `PsMath.h::computeLight`, there is a check if the light requests attenuation. In general, this is done only for directional lights (that are not attenuated), but in this case we know that this point light is not attenuated so passing 0 skips the redundant calculation. In theory this would improve performance, but because this light can only exist alone I don't expect anything measurable.

Reply:
I see 4 tests in `PointLightIlluminationTest` fail due to this change, on MacOS and Windows.
Tests: 
sphereLowerRightPixelColorShouldBeDarkRed
sphereUpperRightPixelColorShouldBeDarkRed
sphereUpperLeftPixelColorShouldBeDarkRed
sphereLowerLeftPixelColorShouldBeDarkRed

Error:
expected:rgba(139,0,0,255) but was:rgba(180,0,0,255)

The tests pass if we pass 1.

Reply:
I'll have a look. What command did you run this test?

Can you also test the point light attenuation under `tests\system\src\test\java\test\javafx\scene\lighting3D\PointLightAttenuationTest.java`? (I think there's a bit of a mess in the tests folders.)

Reply:
Command to run the system test:
`gradle -PFULL_TEST=true -PUSE_ROBOT=true systemTests:test --tests test.robot.test3d.PointLightIlluminationTest`

I ran all the tests, only above test failed:
Command to run all the system tests: `gradle -PFULL_TEST=true -PUSE_ROBOT=true systemTests:test`

Reply:
First of all, I found that the mistake was in the shortcut branch of the shader: it was using the light direction instead of the vector to the light (incident ray), so in the code I need to replace `dot(n, -lightDir)` with `dot(n, -l)`, like is done in the non-shortcut branch. Then the tests passes with the `0` input for no-attenuation. Thanks.

Then I looked at the computation some more and I found something in the computation of the specular component. According to the theory, both in online sources and in the `PhongMaterial` class doc, the computation should be:
`R . V` where `V` is the vector to the eye/cam and `R` is the reflection vector computed by `R = 2(N . L)N - L`, or using the `reflect`  HLSL function, `R = -reflect(N, L)`. So the shader files should look something like this:
```
float3 refl = reflect(l, n);
s = ... dot(-refl, e); // e is the vector to the eye, like V
```
However, looking at the shader files, [already from legacy](https://github.com/openjdk/jfx/blob/c420248b9b459efcfbd3657170d9be0b96b5fb38/modules/javafx.graphics/src/main/native-prism-d3d/hlsl/psMath.h), the vectors are switched:
```
float3 refl = reflect(e, n);
s = ... dot(-refl, l);
```
It looks like the specular computation was wrong from the start.

I tested the visuals on the `master` branch before and after swapping the `l` and `e` vectors and I see no difference in the specular reflection. Rather odd. Will need to look into this more.

The same mistakes are coded into the glsl shaders too, so I should fix the one you found at the very least.

Reply:
@kevinrushforth maybe you can take a look at this too.

Reply:
I would recommend to make only cleanup changes in this PR. So keep `isAttenuated` as 1, and file separate issue to handle the corrections you found.